File: Packages/StrandAnalytics/Sources/StrandAnalytics/TestModeRegistry.swift
import Foundation

/// Whether a guided capture counts nights (Sleep) or days (Battery).
public enum CaptureUnit: String, Sendable, Codable { case nights, days }

/// How a mode captures: a plain on/off toggle, or a guided "wear it for N nights/days" window.
public enum CaptureKind: Sendable, Codable, Equatable {
    case toggle
    case guided(unit: CaptureUnit, defaultCount: Int)   // "defaultCount" (not "default", a reserved word)
}

/// Display priority on the Test Centre screen.
public enum TestPriority: String, Sendable, Codable { case high, med, low }

/// One questionnaire prompt declared by a mode. Answers are stored in meta.json under `id`.
public struct Question: Sendable, Codable, Equatable {
    public let id: String          // stable key stored in the meta.json questionnaire map
    public let prompt: String
    public enum Kind: String, Sendable, Codable { case yesNo, text, time, choice }
    public let kind: Kind
    public let choices: [String]   // only for .choice; else []
    public init(id: String, prompt: String, kind: Kind, choices: [String] = []) {
        self.id = id; self.prompt = prompt; self.kind = kind; self.choices = choices
    }
}

/// A test mode is DATA, not code (spec section 3.1). The screen, the export and the questionnaire all
/// render from this. `captures` / `liveReadout` are declarative ids; the emitters and readout panels
/// bind to them by name. Phase 1 registers exactly `.sleep` and `.battery`.
public struct TestMode: Sendable, Identifiable {
    public let domain: TestDomain
    public let title: String
    public let blurb: String
    public let icon: String                 // SF Symbol on Apple; mapped to a drawable id on Android
    public let priority: TestPriority
    public let captures: [String]           // LogVariable ids, declarative
    public let questionnaire: [Question]
    public let liveReadout: [String]        // ReadoutSpec ids the in-app panel binds
    public let capture: CaptureKind
    public let includesScreenshot: Bool
    public let requires5MG: Bool
    public var id: String { domain.id }

    public init(domain: TestDomain, title: String, blurb: String, icon: String, priority: TestPriority,
                captures: [String], questionnaire: [Question], liveReadout: [String],
                capture: CaptureKind, includesScreenshot: Bool, requires5MG: Bool) {
        self.domain = domain; self.title = title; self.blurb = blurb; self.icon = icon
        self.priority = priority; self.captures = captures; self.questionnaire = questionnaire
        self.liveReadout = liveReadout; self.capture = capture
        self.includesScreenshot = includesScreenshot; self.requires5MG = requires5MG
    }
}

/// The single source the Test Centre IA iterates. Order is priority order on screen. The Kotlin twin is
/// TestModeRegistry.kt, byte-aligned (same ids, titles, captures), verified by a parity test.
public enum TestModeRegistry {

    /// Phase 1 shipped sleep + battery; Phase 2 appends the 🔴 high-pain domains plus the scoring chain
    /// (connection, workouts, display, import, steps, recovery, hrv). Order is screen priority order.
    public static let all: [TestMode] = [
        sleep, connection, workouts, display, dataImport, steps, battery, recovery, hrv,
    ]

    public static func mode(_ d: TestDomain) -> TestMode? { all.first { $0.domain == d } }

    static let sleep = TestMode(
        domain: .sleep, title: "Sleep & Rest",
        blurb: "Wear it a few nights so we can see which gate kept or dropped each sleep run.",
        icon: "bed.double.fill", priority: .high,
        // Only the captures the sleep sink actually receives are kept: the SleepStager gate-verdict ladder
        // (SleepStagerTrace.runLine) plus the RestScorer.subScoreLine composite, i.e. gateTrace, wristOff
        // (the offWristFrac field of the offWrist gate) and restSubScores. Dropped, all unemitted, same
        // rationale Import used for firstFailingRow/failingFileSample/dedupMerge: hrDensity, perEpochFeatures,
        // hypnogramV1V2 (flipLine has no call site), ppgOnlyNight, gravityCoverage (only the sparseBridge
        // gate's sparse= flag, which already rides gateTrace) and skinTempDsp (skin-temp is a Recovery input,
        // never written to the sleep sink). The live-readout keys (hrDensityNow/gravityCoverageNow) are separate.
        captures: ["gateTrace", "wristOff", "restSubScores"],
        questionnaire: [
            Question(id: "sleepTimes", prompt: "Your actual sleep, wake and out-of-bed times?", kind: .text),
            Question(id: "awakeStill", prompt: "Any awake-but-still windows in bed?", kind: .text),
            Question(id: "naps", prompt: "Any naps?", kind: .text),
            Question(id: "shiftWork", prompt: "Shift work or an unusual schedule?", kind: .yesNo),
            Question(id: "chargeTiming", prompt: "When did you charge the strap?", kind: .text),
            Question(id: "healthSleep", prompt: "Is Apple Health / Health Connect also feeding sleep?", kind: .yesNo),
        ],
        liveReadout: ["hrDensityNow", "gravityCoverageNow", "lastNightGateFired"],
        capture: .guided(unit: .nights, defaultCount: 3),
        includesScreenshot: false, requires5MG: false)

    static let connection = TestMode(
        domain: .connection, title: "Connection & Sync",
        blurb: "Turn this on if the strap keeps dropping or won't finish a sync.",
        icon: "antenna.radiowaves.left.and.right", priority: .high,
        captures: ["connectTiming", "bondState", "frameTiming", "reconnectChurn", "offloadProgress",
                   "offloadStalls", "firmwareDecode", "clockDrift", "otherCentral"],
        questionnaire: [
            Question(id: "otherDevicePaired", prompt: "Is another phone or the WHOOP app paired to the strap right now?", kind: .yesNo),
        ],
        liveReadout: ["connectionUptime", "reconnectCount", "lastOffloadResult"],
        capture: .toggle,
        includesScreenshot: false, requires5MG: false)

    static let workouts = TestMode(
        domain: .workouts, title: "Workouts & GPS",
        blurb: "Turn this on if a workout went missing or auto-detect didn't fire.",
        icon: "figure.run", priority: .high,
        captures: ["sessionLifecycle", "hrSamples", "gpsFixes", "autoDetectThresholds",
                   "autoDetectWhy", "crossSourceDedup"],
        questionnaire: [
            Question(id: "startMethod", prompt: "Did you start it manually or expect auto-detect?", kind: .text),
        ],
        liveReadout: ["lastSessionSummary"],
        capture: .toggle,
        includesScreenshot: false, requires5MG: false)

    static let display = TestMode(
        domain: .display, title: "Display & Performance",
        blurb: "Turn this on if a screen looks wrong or feels laggy, then grab a shot.",
        icon: "paintbrush.fill", priority: .high,
        captures: ["screenshot", "deviceMetrics", "frameTimeTrace", "memoryHighWater"],
        questionnaire: [
            Question(id: "screenAndIssue", prompt: "What screen, and what looked or felt wrong (laggy/clipped)?", kind: .text),
        ],
        liveReadout: ["deviceMetricsNow"],
        capture: .toggle,
        includesScreenshot: true, requires5MG: false)

    static let dataImport = TestMode(
        domain: .dataImport, title: "Import & Data Ingest",
        blurb: "Turn this on if a file import dropped rows or came in wrong.",
        icon: "square.and.arrow.down", priority: .high,
        // Only the captures a production import actually emits: parser identity, per-stage rows, reject
        // counts and day deltas. firstFailingRow / failingFileSample / dedupMerge (earlier) and now fileMeta
        // (the import trace emits parser/stage/reject/dayDelta lines but never a file-meta line) were
        // advertised with no emitter on either platform, so they were dropped from BOTH registries to keep
        // the mode honest and the platforms in parity rather than over-promising.
        captures: ["parserVersion", "perStageRows", "rejectCounts", "dayDeltas"],
        questionnaire: [
            Question(id: "appFormatExpected", prompt: "Which app/format, and what did you expect to import?", kind: .text),
        ],
        liveReadout: ["lastImportSummary"],
        capture: .toggle,
        includesScreenshot: false, requires5MG: false)

    static let steps = TestMode(
        domain: .steps, title: "Steps",
        blurb: "Turn this on if your step count looks off versus your phone.",
        icon: "shoeprints.fill", priority: .high,
        captures: ["motionVolume", "stepCalibration", "phoneReferenceCount", "rawStepCounter",
                   "wrapAwareDeltas", "droppedDeltas"],
        questionnaire: [
            Question(id: "otherTrackerSteps", prompt: "What did your phone or another tracker report for the same day?", kind: .text),
        ],
        liveReadout: ["stepsToday", "calibrationState"],
        capture: .toggle,
        includesScreenshot: false, requires5MG: false)

    static let battery = TestMode(
        domain: .battery, title: "Battery & Charging",
        blurb: "Wear it a few days so we can fit your real discharge slope.",
        icon: "battery.50", priority: .med,
        // Dropped offWristGaps: nothing emits an off-wrist-gap line for the battery discharge run.
        captures: ["socSeries", "chargeSteps", "dischargeRun", "fittedSlope",
                   "sourceMeasuredVsRated", "batteryGates"],
        questionnaire: [
            Question(id: "whoopAppInstalled", prompt: "Is the official WHOOP app installed?", kind: .yesNo),
            Question(id: "otherPhonePaired", prompt: "Is another phone paired to the strap?", kind: .yesNo),
            Question(id: "chargedInWindow", prompt: "Did you charge during the capture?", kind: .yesNo),
            Question(id: "batterySaverApps", prompt: "Any battery-saver apps running?", kind: .text),
        ],
        liveReadout: ["currentSoc", "estimateDaysLeft", "slopeSource"],
        capture: .guided(unit: .days, defaultCount: 3),
        includesScreenshot: false, requires5MG: false)

    static let recovery = TestMode(
        domain: .recovery, title: "Recovery (Charge)",
        blurb: "Turn this on if Charge looks wrong, to see which term moved it.",
        icon: "heart.text.square.fill", priority: .med,
        // Dropped forecastInputs: the recovery trace emits baseline/term/nilTerm/renorm/score, no forecast line.
        captures: ["chargeTermBreakdown", "baselinesPerNight", "termZScores", "nilTerm"],
        questionnaire: [
            Question(id: "recalHealthHrv", prompt: "Recent recalibration? Is Apple Health / Health Connect feeding HRV?", kind: .text),
        ],
        liveReadout: ["lastChargeBreakdown"],
        capture: .toggle,
        includesScreenshot: false, requires5MG: false)

    static let hrv = TestMode(
        domain: .hrv, title: "HRV & Autonomic",
        blurb: "Turn this on if HRV reads nil or looks off, to see the clean beats.",
        icon: "waveform.path.ecg", priority: .med,
        // The HRV trace emits the path/nInput/nClean/reject line, the minBeats gate and the
        // "hrv rmssd=... sdnn=... meanNN=..." line. Dropped: respRsa (RSA respiration isn't computed in the
        // HRV path), rawRR (the trace emits the beat COUNT nInput=, never the raw R-R list) and
        // spotVsContinuous (the sole call site hardcodes path="spot", so it never emits "continuous").
        captures: ["nInputCleanRejected", "rmssdSdnn", "minBeatsCleared"],
        questionnaire: [
            Question(id: "otherAppHrv", prompt: "Is another app feeding HRV to Apple Health / Health Connect?", kind: .yesNo),
        ],
        liveReadout: ["lastHrvComputation"],
        capture: .toggle,
        includesScreenshot: false, requires5MG: false)
}

File: android/app/src/main/java/com/noop/testcentre/TestModeRegistry.kt
package com.noop.testcentre

/** Whether a guided capture counts nights (Sleep) or days (Battery). */
enum class CaptureUnit { NIGHTS, DAYS }

/** How a mode captures: a plain toggle, or a guided "wear it N nights/days" window. Twin of the Swift
 *  CaptureKind. */
sealed class CaptureKind {
    object Toggle : CaptureKind()
    data class Guided(val unit: CaptureUnit, val defaultCount: Int) : CaptureKind()
}

/** Display priority on the Test Centre screen. */
enum class TestPriority { HIGH, MED, LOW }

/** One questionnaire prompt declared by a mode. Answers stored in meta.json under [id]. */
data class Question(
    val id: String,
    val prompt: String,
    val kind: Kind,
    val choices: List<String> = emptyList(),
) {
    enum class Kind { YES_NO, TEXT, TIME, CHOICE }
}

/** A test mode is DATA, not code (spec section 3.1). The screen, export and questionnaire all render
 *  from this. Twin of the Swift TestMode, byte-aligned by a parity test. */
data class TestMode(
    val domain: TestDomain,
    val title: String,
    val blurb: String,
    val icon: String,                 // drawable id on Android; SF Symbol on Apple
    val priority: TestPriority,
    val captures: List<String>,
    val questionnaire: List<Question>,
    val liveReadout: List<String>,
    val capture: CaptureKind,
    val includesScreenshot: Boolean,
    val requires5MG: Boolean,
) {
    val id: String get() = domain.id
}

/** The single source the Test Centre IA iterates. Order is priority order. Twin of the Swift
 *  TestModeRegistry; same ids/titles/captures, verified by [TestModeRegistryParityTest]. */
object TestModeRegistry {

    val all: List<TestMode> = listOf(
        sleep(), connection(), workouts(), display(), dataImport(), steps(), battery(), recovery(), hrv(),
    )

    fun mode(d: TestDomain): TestMode? = all.firstOrNull { it.domain == d }

    private fun sleep() = TestMode(
        domain = TestDomain.SLEEP, title = "Sleep & Rest",
        blurb = "Wear it a few nights so we can see which gate kept or dropped each sleep run.",
        icon = "ic_bed", priority = TestPriority.HIGH,
        // Only the captures the .sleep sink actually receives are kept. That sink only ever gets the
        // SleepStager gate-verdict ladder (SleepStagerTrace.runLine) plus the RestScorer.subScoreLine
        // composite, so gateTrace, wristOff (the offWristFrac field of the offWrist gate) and restSubScores
        // are the whole set. Dropped, all unemitted, same rationale the Import mode used for
        // firstFailingRow/failingFileSample/dedupMerge: hrDensity, perEpochFeatures, hypnogramV1V2 (flipLine
        // has no call site), ppgOnlyNight, gravityCoverage (only the sparseBridge gate's sparse= flag, which
        // already rides gateTrace) and skinTempDsp (skin-temp is a Recovery input, never written to the
        // sleep sink). The live-readout keys (hrDensityNow/gravityCoverageNow) are separate and unchanged.
        captures = listOf("gateTrace", "wristOff", "restSubScores"),
        questionnaire = listOf(
            Question("sleepTimes", "Your actual sleep, wake and out-of-bed times?", Question.Kind.TEXT),
            Question("awakeStill", "Any awake-but-still windows in bed?", Question.Kind.TEXT),
            Question("naps", "Any naps?", Question.Kind.TEXT),
            Question("shiftWork", "Shift work or an unusual schedule?", Question.Kind.YES_NO),
            Question("chargeTiming", "When did you charge the strap?", Question.Kind.TEXT),
            Question("healthSleep", "Is Apple Health / Health Connect also feeding sleep?", Question.Kind.YES_NO),
        ),
        liveReadout = listOf("hrDensityNow", "gravityCoverageNow", "lastNightGateFired"),
        capture = CaptureKind.Guided(CaptureUnit.NIGHTS, 3),
        includesScreenshot = false, requires5MG = false,
    )

    private fun connection() = TestMode(
        domain = TestDomain.CONNECTION, title = "Connection & Sync",
        blurb = "Turn this on if the strap keeps dropping or won't finish a sync.",
        icon = "ic_antenna", priority = TestPriority.HIGH,
        captures = listOf("connectTiming", "bondState", "frameTiming", "reconnectChurn", "offloadProgress",
            "offloadStalls", "firmwareDecode", "clockDrift", "otherCentral"),
        questionnaire = listOf(
            Question("otherDevicePaired", "Is another phone or the WHOOP app paired to the strap right now?", Question.Kind.YES_NO),
        ),
        liveReadout = listOf("connectionUptime", "reconnectCount", "lastOffloadResult"),
        capture = CaptureKind.Toggle,
        includesScreenshot = false, requires5MG = false,
    )

    private fun workouts() = TestMode(
        domain = TestDomain.WORKOUTS, title = "Workouts & GPS",
        blurb = "Turn this on if a workout went missing or auto-detect didn't fire.",
        icon = "ic_run", priority = TestPriority.HIGH,
        captures = listOf("sessionLifecycle", "hrSamples", "gpsFixes", "autoDetectThresholds",
            "autoDetectWhy", "crossSourceDedup"),
        questionnaire = listOf(
            Question("startMethod", "Did you start it manually or expect auto-detect?", Question.Kind.TEXT),
        ),
        liveReadout = listOf("lastSessionSummary"),
        capture = CaptureKind.Toggle,
        includesScreenshot = false, requires5MG = false,
    )

    private fun display() = TestMode(
        domain = TestDomain.DISPLAY, title = "Display & Performance",
        blurb = "Turn this on if a screen looks wrong or feels laggy, then grab a shot.",
        icon = "ic_paintbrush", priority = TestPriority.HIGH,
        captures = listOf("screenshot", "deviceMetrics", "frameTimeTrace", "memoryHighWater"),
        questionnaire = listOf(
            Question("screenAndIssue", "What screen, and what looked or felt wrong (laggy/clipped)?", Question.Kind.TEXT),
        ),
        liveReadout = listOf("deviceMetricsNow"),
        capture = CaptureKind.Toggle,
        includesScreenshot = true, requires5MG = false,
    )

    private fun dataImport() = TestMode(
        domain = TestDomain.IMPORT, title = "Import & Data Ingest",
        blurb = "Turn this on if a file import dropped rows or came in wrong.",
        icon = "ic_import", priority = TestPriority.HIGH,
        // Only the captures a production import actually emits (parser identity, per-stage rows, reject
        // counts, day deltas). firstFailingRow / failingFileSample / dedupMerge (earlier) and now fileMeta
        // (the import trace emits parser/stage/reject/dayDelta lines but never a file-meta line) were
        // advertised with no emitter on either platform, so they are dropped from BOTH registries (parity)
        // rather than over-promising the mode.
        captures = listOf("parserVersion", "perStageRows", "rejectCounts", "dayDeltas"),
        questionnaire = listOf(
            Question("appFormatExpected", "Which app/format, and what did you expect to import?", Question.Kind.TEXT),
        ),
        liveReadout = listOf("lastImportSummary"),
        capture = CaptureKind.Toggle,
        includesScreenshot = false, requires5MG = false,
    )

    private fun steps() = TestMode(
        domain = TestDomain.STEPS, title = "Steps",
        blurb = "Turn this on if your step count looks off versus your phone.",
        icon = "ic_steps", priority = TestPriority.HIGH,
        captures = listOf("motionVolume", "stepCalibration", "phoneReferenceCount", "rawStepCounter",
            "wrapAwareDeltas", "droppedDeltas"),
        questionnaire = listOf(
            Question("otherTrackerSteps", "What did your phone or another tracker report for the same day?", Question.Kind.TEXT),
        ),
        liveReadout = listOf("stepsToday", "calibrationState"),
        capture = CaptureKind.Toggle,
        includesScreenshot = false, requires5MG = false,
    )

    private fun battery() = TestMode(
        domain = TestDomain.BATTERY, title = "Battery & Charging",
        blurb = "Wear it a few days so we can fit your real discharge slope.",
        icon = "ic_battery", priority = TestPriority.MED,
        // Dropped offWristGaps: nothing emits an off-wrist-gap line for the battery discharge run.
        captures = listOf("socSeries", "chargeSteps", "dischargeRun", "fittedSlope",
            "sourceMeasuredVsRated", "batteryGates"),
        questionnaire = listOf(
            Question("whoopAppInstalled", "Is the official WHOOP app installed?", Question.Kind.YES_NO),
            Question("otherPhonePaired", "Is another phone paired to the strap?", Question.Kind.YES_NO),
            Question("chargedInWindow", "Did you charge during the capture?", Question.Kind.YES_NO),
            Question("batterySaverApps", "Any battery-saver apps running?", Question.Kind.TEXT),
        ),
        liveReadout = listOf("currentSoc", "estimateDaysLeft", "slopeSource"),
        capture = CaptureKind.Guided(CaptureUnit.DAYS, 3),
        includesScreenshot = false, requires5MG = false,
    )

    private fun recovery() = TestMode(
        domain = TestDomain.RECOVERY, title = "Recovery (Charge)",
        blurb = "Turn this on if Charge looks wrong, to see which term moved it.",
        icon = "ic_recovery", priority = TestPriority.MED,
        // Dropped forecastInputs: the recovery trace emits baseline/term/nilTerm/renorm/score, no forecast line.
        captures = listOf("chargeTermBreakdown", "baselinesPerNight", "termZScores", "nilTerm"),
        questionnaire = listOf(
            Question("recalHealthHrv", "Recent recalibration? Is Apple Health / Health Connect feeding HRV?", Question.Kind.TEXT),
        ),
        liveReadout = listOf("lastChargeBreakdown"),
        capture = CaptureKind.Toggle,
        includesScreenshot = false, requires5MG = false,
    )

    private fun hrv() = TestMode(
        domain = TestDomain.HRV, title = "HRV & Autonomic",
        blurb = "Turn this on if HRV reads nil or looks off, to see the clean beats.",
        icon = "ic_hrv", priority = TestPriority.MED,
        // The HRV trace (HrvAnalyzerTrace) emits the path/nInput/nClean/reject line, the minBeats gate and
        // the "hrv rmssd=... sdnn=... meanNN=..." line. Dropped: respRsa (RSA respiration isn't computed in
        // the HRV path), rawRR (the trace emits the beat COUNT nInput=, never the raw R-R list) and
        // spotVsContinuous (the sole call site hardcodes path="spot", so it never emits "continuous").
        captures = listOf("nInputCleanRejected", "rmssdSdnn", "minBeatsCleared"),
        questionnaire = listOf(
            Question("otherAppHrv", "Is another app feeding HRV to Apple Health / Health Connect?", Question.Kind.YES_NO),
        ),
        liveReadout = listOf("lastHrvComputation"),
        capture = CaptureKind.Toggle,
        includesScreenshot = false, requires5MG = false,
    )
}
